File: serve.js
// daemon.js
var fs = require('fs');
var url = require('url');
var http = require('http');
var querystring = require('querystring');

var VERSION = 17;

function handleRequest(request, response) {
  function acceptWithJSON(json) {
    response.writeHead(200, {
      'Access-Control-Allow-Origin': origin,
      'Content-Type': 'application/json',
    });
    response.end(JSON.stringify(json));
  }

  function acceptWithBinary(buffer) {
    response.writeHead(200, {
      'Access-Control-Allow-Origin': origin,
      'Content-Type': 'application/octet-stream',
    });
    response.end(buffer);
  }

  function ignore(comment) {
    console.log('ignoring: ' + comment);
    response.writeHead(404, { 'Content-Type': 'text/plain' });
    response.end();
  }

  console.log(request.method + ' ' + request.url);
  if (!fonts) {
    return ignore('ignoring request before font data is present');
  }

  // Don't expose this API to everyone
  if (request.method !== 'GET') {
    return ignore('invalid method: ' + request.method);
  }

  var requestURL = url.parse(request.url);
  if (!requestURL) {
    return ignore();
  }

  // Note: Origin can contain multiple whitespace separated origins due to
  // redirects, but that shouldn't happen for us. If it does happen, the
  // isValidOrigin check below will fail.
  var origin = request.headers.origin;
  origin = '*';

  // Whitelist certain URLs
  switch (requestURL.pathname) {
    case '/figma/version': {
      return acceptWithJSON({ version: VERSION });
    }

    case '/figma/font-files': {
      return acceptWithJSON({ version: VERSION, fontFiles: fonts });
    }

    case '/figma/update': {
      var query = querystring.parse(requestURL.query);
      if (!query || !query.version) {
        return ignore('missing version');
      }
      if (+query.version === (query.version | 0)) {
        if (+query.version > VERSION) {
          console.log('CHECK FOR UPDATES'); // The parent process is monitoring our stdout and will see this
        }
        return acceptWithJSON({ version: VERSION });
      }
      return ignore('invalid version: ' + query.version);
    }

    // Whitelist served files
    case '/figma/font-file': {
      var query = querystring.parse(requestURL.query);
      if (!query || !query.file) {
        return ignore('missing file');
      }
      if (fonts.hasOwnProperty(query.file)) {
        try {
          var buffer = fs.readFileSync(query.file);
          if (buffer.length > 0) {
            return acceptWithBinary(buffer);
          }
        } catch (e) {
        }
      }
      return ignore('invalid file: ' + query.file);
    }

    default: {
      return ignore('invalid url: ' + requestURL.pathname);
    }
  }
}

function startServer(line) {
  console.log('starting server');

  const initServer = (server, port) => {
    const listen = () => server.listen(port, 'localhost');

    server.on('error', error => {
      if (error.code === 'EADDRINUSE') {
        // Try again after a short delay.
        setTimeout(listen, 1000 * 10);
      } else {
        throw error;
      }
    });

    listen();
  };

  // HTTP for browsers that support 127.0.0.1 without triggering the mixed
  // content blocker.
  httpServer = http.createServer(handleRequest);
  initServer(httpServer, 44950);
}

var fonts = null;
var httpServer = null;

function scanfontsjs(roots) {
    // scanfonts.js
    const fs = require('fs')
    const path = require('path')
    const zlib = require('zlib')

    function parse_font_file(fn) {
        let data = fs.readFileSync(fn)
        let spec;
        switch (data.readUInt32BE(0)) {
            case 0x74746366: // ttcf
                spec = analyze_font(data, otc_reader); break
            case 0x774f4646: // wOFF
                spec = analyze_font(data, woff_reader); break
            case 0x774f4632: // wOF2
                spec = analyze_font(data, woff2_reader); break
            default:
                spec = analyze_font(data, otf_reader); break
        }
        return Array.from(spec)
    }

    function* analyze_font(data, reader) {
        const tags = new Set(['head', 'name', 'OS/2'])
        for (let section of reader(data, tags)) {
            yield extract_attributes(section)
        }
    }

    function extract_attributes(section) {
        function parse_head(data, attribs) {
            const major = data.readUInt16BE(0)
            const minor = data.readUInt16BE(2)
            if (major !== 1 || minor !== 0) {
                throw new Error(`Unsupported format version (${major},${minor})`)
            }
            const magic = data.readUInt32BE(12)
            if (magic !== 0x5F0F3CF5) {
                throw new Error(`Invalid signature ${magic}`)
            }
            const macStyle = data.readUInt16BE(44)
            attribs['italic'] = ((macStyle & 2) !== 0)
        }

        function parse_name(data, attribs) {
            const nameids = new Map([[1, 'family'], [2, 'style'], [6, 'postscript']])
            const version = data.readUInt16BE(0)
            const nrecs = data.readUInt16BE(2)
            const strings = data.readUInt16BE(4)
            for (let off = 6, i = 0; i < nrecs; ++i, off += 12) {
                const nameid = data.readUInt16BE(off+6)
                const sn = nameids.get(nameid)
                if (sn) {
                    const splat = data.readUInt16BE(off+0)
                    const sspec = data.readUInt16BE(off+2)
                    const slang = data.readUInt16BE(off+4)
                    const len = data.readUInt16BE(off+8)
                    const soff = data.readUInt16BE(off+10)
                    const s = decodeString(data.subarray(strings+soff, strings+soff+len), splat, sspec, slang)
                    if (s) {
                        attribs[sn] = s
                    }
                }
            }
        }

        function parse_OS_2(data, attribs) {
            attribs['weight'] = data.readUInt16BE(4)
            attribs['width'] = data.readUInt16BE(6)
        }

        let attribs = {}
        for (let [tag, body] of section) {
            switch (tag) {
                case 'head': parse_head(body, attribs); break
                case 'name': parse_name(body, attribs); break
                case 'OS/2': parse_OS_2(body, attribs); break
            }
        }
        return attribs
    }

    function* otc_reader(data, tags) {
        const major = data.readUInt16BE(4)
        const minor = data.readUInt16BE(6)
        if ((major !== 1 && major !== 2) || (minor !== 0)) {
            throw new Error(`Unsupported format version (${major},${minor})`)
        }
        const nfonts = data.readUInt32BE(8)
        for (let i = 0, off = 12; i < nfonts; ++i, off += 4) {
            const p = data.readUInt32BE(off)
            yield* otf_reader(data, tags, p)
        }
    }

    function* otf_reader(data, tags, tableOffset=0) {
        function* inner() {
            const sig = data.readUInt32BE(tableOffset+0)
            if (!(sig === 0x4F54544F || sig === 0x00010000 || sig === 0x74727565)) { // OTTO, 0100, true
                throw new Error(`Unsupported format ${sig}`)
            }
            const ntables = data.readUInt16BE(tableOffset+4)
            for (let i = 0, off = tableOffset+12; i < ntables; ++i, off += 16) {
                const tag = data.subarray(off, off+4).toString('utf-8')
                if (tags.has(tag)) {
                    const p = data.readUInt32BE(off+8)
                    const n = data.readUInt32BE(off+12)
                    const body = data.subarray(p, p+n)
                    yield [tag, body]
                }
            }
        }
        yield inner()
    }

    function* woff_reader(data, tags) {
        function* inner() {
            const ntables = data.readUInt16BE(12)
            for (let i = 0, off = 44; i < ntables; ++i, off += 20) {
                const tag = data.subarray(off, off+4).toString('utf-8')
                if (tags.has(tag)) {
                    const p = data.readUInt32BE(off+4)
                    const n = data.readUInt32BE(off+8)
                    const body = zlib_decompress(data.subarray(p, p+n))
                    yield [tag, body]
                }
            }
        }
        yield inner()
    }

    function* woff2_reader(data, tags) {
        const flavor = data.readUInt32BE(4)
        const ntables = data.readUInt16BE(12)
        const compressed_size = data.readUInt32BE(20)
        const known_tags = new Map([[1, 'head'], [5, 'name'], [6, 'OS/2']])
        let offsets = []
        let off = 48
        let p = 0, n
        for (let i = 0; i < ntables; ++i) {
            const flags = data[off]
            off += 1
            let tag = flags & 0x3f
            const trver = flags >> 6
            const transformed = (tag === 10 || tag === 11) ? (trver !== 3) : (trver !== 0)
            if (tag === 63) {
                tag = data.subarray(off, off+4).toString('utf8')
                off += 4
            }
            else {
                tag = known_tags.get(tag)
            }
            [n, off] = read_UIntBase128(data, off)
            if (transformed) {
                [n, off] = read_UIntBase128(data, off)
            }
            offsets.push([tag, p, n])
            p += n
        }

        let directory = [offsets]
        if (flavor === 0x74746366) { // ttcf
            off += 4
            directory = []
            let nfonts = 0;
            [nfonts, off] = read_255UInt16(data, off)
            for (let fi = 0; fi < nfonts; ++fi) {
                let ntables = 0;
                [ntables, off] = read_255UInt16(data, off)
                off += 4
                let xs = []
                for (let ti = 0; ti < ntables; ++ti) {
                    let idx = 0;
                    [idx, off] = read_255UInt16(data, off)
                    xs.push(offsets[idx])
                }
                directory.push(xs)
            }
        }

        const compressed = data.subarray(off, off+compressed_size)
        data = zlib.brotliDecompressSync(compressed)

        function* entry(offsets) {
            for (let [tag, p, n] of offsets) {
                if (tags.has(tag)) {
                    const body = data.subarray(p, p+n)
                    yield [tag, body]
                }
            }
        }

        for (let t of directory) {
            yield entry(t)
        }
    }

    function read_255UInt16(data, off) {
        let n = data[off]
        switch (n) {
            case 253:
                n = (data[off+1] << 8) | data[off+2]
                off += 3; break
            case 254:
                n = 506 + data[off+1]
                off += 2; break
            case 255:
                n = 253 + data[off+1]
                off += 2; break
            default:
                off += 1; break
        }
        return [n, off]
    }

    function read_UIntBase128(data, off) {
        for (let n = 0, i = 0; i < 5; ++i) {
            const x = data[off]
            off += 1
            n = (n << 7) | (x & 0x7f)
            if ((x & 0x80) == 0) {
                return [n, off]
            }
        }
        throw new Error('Invalid UIntBase128 value')
    }

    function zlib_decompress(data) {
        try {
            return zlib.inflateSync(data)
        }
        catch {}
        return data
    }

    function decodeString(s, platform, spec, lang) {
        switch (platform) {
            case 0: return s.toString('utf8')
            case 1:
                if (spec === 0 && lang === 0) {
                    return s.toString('utf8')
                }
                break
            case 3:
                if (spec === 1 && (lang % 256 === 9)) {
                    return s.swap16().toString('utf16le')
                }
        }
    }

    function* glob(roots, rx) {
        let fringe = [...roots]
        while (fringe.length) {
            let dir = fringe.pop()
            let subdirs;
            try {
                subdirs = fs.readdirSync(dir, {withFileTypes:true})
            } catch { continue }
            for (let f of subdirs) {
                if (f.name.match(/^\./)) { continue }
                let fullname = path.join(dir, f.name)
                if (f.isFile() && rx.test(f.name)) {
                    yield fullname
                }
                else if (f.isDirectory()) {
                    fringe.push(fullname)
                }
            }
        }
    }

    function* enumerate_fonts(roots) {
        for (let fn of glob(roots, /\.(?:ttf|ttc|otf|otc|woff2?)$/i)) {
            try {
                let spec = parse_font_file(fn)
                yield [fn, spec]
            } catch(e) { console.error(fn); console.error(e) }
        }
    }

    return Object.fromEntries(enumerate_fonts(roots))
}

var process = require('process');
let roots = process.argv.slice(2)
if (!roots.length) {
    roots = ['.']
}
fonts = scanfontsjs(roots)

startServer()

File: serve.py
#!/usr/bin/env python
import http.server
import json
import re
import socketserver
import struct
import sys
import traceback
import zlib
from pathlib import Path
from urllib.parse import urlsplit, parse_qs


VERSION = 17
BIND_PORT = 44950
FONTS = None


class RequestHandler (http.server.BaseHTTPRequestHandler):

    def _accept_json(self, obj):
        s = json.dumps(obj, separators=',:').encode('utf-8')
        self._accept(s, 'application/json')

    def _accept_binary(self, data):
        self._accept(data, 'application/octet-stream')

    def _accept(self, data, content_type):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Content-Type', content_type)
        self.send_header('Content-Length', len(data))
        self.end_headers()
        self.wfile.write(data)

    def _ignore(self, comment):
        print(f'ignoring: {comment}', file=sys.stderr)
        self.send_response(404)
        self.send_header('Content-Type', 'text/plain')
        self.end_headers()

    def do_GET(self):
        _,_,path,query,_ = urlsplit(self.path)

        if path == '/figma/version':
            self._accept_json({'version': VERSION})

        elif path == '/figma/font-files':
            self._accept_json({'version': VERSION, 'fontFiles': FONTS})

        elif path == '/figma/update':
            params = parse_qs(query)
            ver = params.get('version')
            if ver is None:
                self._ignore('missing version')
            else:
                try:
                    ver, = ver
                    v = float(ver)
                    if v > VERSION:
                        print('CHECK FOR UPDATES', file=sys.stderr)
                    self._accept_json({'version': VERSION})
                except (TypeError, ValueError):
                    self._ignore(f'invalid version: {ver}')

        elif path == '/figma/font-file':
            params = parse_qs(query)
            file = params.get('file')
            if file is None:
                self._ignore('missing file')
            try:
                file, = file
                if file in FONTS:
                    with open(file, 'rb') as fp:
                        data = fp.read()
                    self._accept_binary(data)
                    return
            except:
                print(repr(file), file=sys.stderr)
                traceback.print_exc()
            self._ignore(f'invalid file: {file}')

        else:
            self._ignore(f'invalid url: {self.path}')


def parse_font_file(fn):
    with open(fn, 'rb') as fp:
        data = fp.read()
    sig = data[:4]
    if sig == b'ttcf':
        spec = analyze_font(data, otc_reader)
    elif sig == b'wOFF':
        spec = analyze_font(data, woff_reader)
    elif sig == b'wOF2':
        spec = analyze_font(data, woff2_reader)
    else:
        spec = analyze_font(data, otf_reader)
    return list(spec)


def analyze_font(data, reader):
    for section in reader(data, [b'head', b'name', b'OS/2']):
        yield extract_attributes(section)


def extract_attributes(section):
    attribs = dict()
    for tag, body in section:

        if tag == b'head':
            major, minor = struct.unpack('>HH', body[:4])
            if not ((major == 1) and (minor == 0)):
                raise Exception(f'Unsupported format version {(major, minor)}')

            sig, = struct.unpack('>I', body[12:16])
            if sig != 0x5F0F3CF5:
                raise Exception(f'Invalid signature {sig:08X}')

            mac_style, = struct.unpack('>H', body[44:46])
            italic = (mac_style & 2) != 0
            attribs['italic'] = italic

        elif tag == b'name':
            nameids = {1: 'family', 2: 'style', 6: 'postscript'}
            version, nrecs, table = struct.unpack('>HHH', body[:6])
            for i in range(nrecs):
                off = 6 + i * 12
                splat, sspec, slang, nameid, slen, soff = struct.unpack('>HHHHHH', body[off:off+12])
                soff += table
                sn = nameids.get(nameid)
                if sn:
                    s = decode_string(body[soff:soff+slen], splat, sspec, slang)
                    if s:
                        attribs[sn] = s

        elif tag == b'OS/2':
            weight, width = struct.unpack('>HH', body[4:8])
            attribs['weight'] = weight
            attribs['width'] = width

    return attribs


def otc_reader(data, tags):
    major, minor, nfonts = struct.unpack('>HHI', data[4:12])
    if not ((major == 1 or major == 2) and (minor == 0)):
        raise Exception(f'Unsupported format version {(major, minor)}')
    for i in range(nfonts):
        off = 12 + i * 4
        p, = struct.unpack('>I', data[off:off+4])
        yield from otf_reader(data, tags, p)


def otf_reader(data, tags, table_offset=0):
    def inner():
        off = table_offset + 0
        sig = data[off:off+4]
        if sig not in (b'OTTO', b'\x00\x01\x00\x00', b'true'):
            raise Exception(f'Unsupported format {sig}')

        off = table_offset + 4
        ntables, = struct.unpack('>H', data[off:off+2])
        for i in range(ntables):
            off = table_offset + 12 + i * 16
            tag = data[off:off+4]
            if tag in tags:
                p,n = struct.unpack('>II', data[off+8:off+16])
                body = data[p:p+n]
                yield (tag, body)

    yield inner()


def woff_reader(data, tags):
    def inner():
        ntables, = struct.unpack('>H', data[12:14])
        for i in range(ntables):
            off = 44 + i * 20
            tag = data[off:off+4]
            if tag in tags:
                p,n = struct.unpack('>II', data[off+4:off+12])
                body = zlib_decompress(data[p:p+n])
                yield (tag, body)

    yield inner()


def woff2_reader(data, tags):
    import brotli
    flavor = data[4:8]
    ntables, = struct.unpack('>H', data[12:14])
    compressed_size, = struct.unpack('>I', data[20:24])
    known_tags = {1: b'head', 5: b'name', 6: b'OS/2'}
    offsets = list()
    off = 48
    p = 0
    for i in range(ntables):
        flags = data[off]
        off += 1
        tag = flags & 0x3f
        trver = flags >> 6
        transformed = (trver != 3) if (tag == 10 or tag == 11) else (trver != 0)
        if tag == 63:
            stag = data[off:off+4]
            off += 4
        else:
            stag = known_tags.get(tag)
        n, off = read_UIntBase128(data, off)
        if transformed:
            n, off = read_UIntBase128(data, off)
        offsets.append((stag, p, n))
        p += n

    directory = [offsets]
    if flavor == b'ttcf':
        off += 4
        directory = list()
        nfonts, off = read_255UInt16(data, off)
        for _ in range(nfonts):
            ntables, off = read_255UInt16(data, off)
            flavor = data[off:off+4]
            off += 4
            xs = list()
            for _ in range(ntables):
                idx, off = read_255UInt16(data, off)
                xs.append(offsets[idx])
            directory.append(xs)

    compressed = data[off:off+compressed_size]
    data = brotli.decompress(compressed)

    def entry(offsets):
        for tag, p, n in offsets:
            if tag in tags:
                body = data[p:p+n]
                yield (tag, body)

    for t in directory:
        yield entry(t)


def read_255UInt16(data, off):
    x = data[off]
    n = x
    if x == 253:
        n = (data[off+1] << 8) | data[off+2]
        off += 3
    elif x == 254:
        n = 506 + data[off+1]
        off += 2
    elif x == 255:
        n = 253 + data[off+1]
        off += 2
    else:
        off += 1
    return n, off


def read_UIntBase128(data, off):
    n = 0
    for _ in range(5):
        x = data[off]
        off += 1
        n = (n << 7) | (x & 0x7f)
        if not (x & 0x80):
            return n, off
    raise Exception('Invalid UIntBase128 value')


def zlib_decompress(data):
    try:
        return zlib.decompress(data)
    except zlib.error:
        pass
    return data


def decode_string(s, platform, spec, lang):
    if platform == 0:
        return s.decode('utf-8')
    elif platform == 1:
        if spec == 0 and lang == 0:
            return s.decode('utf-8')
    elif platform == 3:
        if spec == 1 and (lang % 256) == 9:
            return s.decode('utf-16be')


def scan_fonts(roots):
    def glob(roots, rx):
        fringe = [Path(s) for s in roots]
        while fringe:
            d = fringe.pop()
            for fn in d.iterdir():
                if fn.name.startswith('.'): continue
                if fn.is_file() and rx.search(fn.name):
                    yield fn
                elif fn.is_dir():
                    fringe.append(fn)

    def enumerate_fonts(roots):
        rx = re.compile(r'\.(ttf|ttc|otf|otc|woff2?)$', re.I)
        for fn in glob(roots, rx):
            try:
                spec = parse_font_file(fn)
                yield [str(fn), spec]
            except:
                print(fn, file=sys.stderr)
                traceback.print_exc()

    return dict(enumerate_fonts(roots))


def start_server():
    with socketserver.TCPServer(('localhost', BIND_PORT), RequestHandler) as httpd:
        print('starting server')
        httpd.serve_forever()


if __name__ == '__main__':
    import argparse
    parser = argparse.ArgumentParser(description='Serve local fonts to Figma.')
    parser.add_argument('roots', metavar='dir', nargs='*', default=['.'], help='a directory with font files')
    args = parser.parse_args()

    FONTS = scan_fonts(args.roots)
    start_server()
